### ISSUE TITLE
In a keyboard multi layout situation: optionally show the current layout in SSD

### ISSUE THREAD

**Issue by Consolatis** [labwc/labwc#1081]
This somewhat works around the missing support for a panel to show the indicator.

Although we internally only have the indices and a string like "English" or "German" in the compiled keymaps we could parse the `XKB_DEFAULT_LAYOUT` env var and translate the index into the short form. All credits to @johanmalm for the whole SSD keyboard indicator idea.

**Comment by kyak** [comment]
@Consolatis for the keyboard layout per window situation, what should be displayed in the OSD?

**Comment by Consolatis** [comment]
I don't think we should display anything in the `OSD` here (e.g. the `Alt-Tab` window).

For the `SSD` I am not sure, there were multiple ideas floating around, just something like `[en]` after or before the title or `[en]` right next to the `menu` button on the left. Thoughts @johanmalm ?

**Comment by johanmalm** [comment]
Yes, exactly my thoughts.
Could be part of titleLayout when we get to that.

**Comment by stefonarch** [comment]
My 2 cents: I always use an indicator on the panel, some windows have no title bar.
If displayed there it should be optional (and beside the menu is a good place IMO).

**Comment by kyak** [comment]
Most of my window lack SSD (title bar) to save screen real estate. Also, it is typical for a user to expect layout indicator in the same place every time (e.g. in the panel) - it is easier to take a glance at the same place.

If you are in principal not against drawing layout indicator in the title bar, maybe we should expand on this idea?

For example, labwc could display a small floating window in the user-defined location. Basically labwc would ship keyboard layout indicator (but then drawing layout in SSD is also kind of shipping layout indicator). Is it a bad idea?

**Comment by Consolatis** [comment]
Well, we could also just modify the title itself rather than the SSD. Then a panel would automatically show the indicator in front of every application.

**Comment by stefonarch** [comment]
I don't think that this should be by some window but it's basically a panel thing as it always has been.

I'd rather like some kind of api which can be used by custom scripts/panels instead of this workaround with LED states which also works only with 2 layouts. Or just a file in /tmp which could be read, in lxqt-panel the custom command plugin can also display flags now:

![screen_area_dom_11:54:11_](https://github.com/labwc/labwc/assets/10681413/1bf8e3c4-bfdb-4432-b872-21394116da38)

**Comment by kyak** [comment]
> Well, we could also just modify the title itself rather than the SSD. Then a panel would automatically show the indicator in front of (or after) every application, same for the SSD. (Disabled by default obviously.)
> 

Items in panel tend to get smaller with the number of windows. Adding more text there steals valuable space.

Also, this is again spreading layout indicator all over the place instead of keeping it in the one and only place.

> Edit: Another random idea, we could react to the (keyboard) focus switch to a window and draw a small overlay with the current layout somewhere on top of the window which gets hidden after some timeout.

Good idea. This could also be drawn in the middle of the screen, just like window switcher OSD.
